### DISCUSSION TITLE
Is Java 17 supported?

### DISCUSSION THREAD

**Discussion by Ratgor** [jpype-project/jpype#1138] [Q&A] [answered]
Does jpype have official support for Java 17?
When it is planned to be added, if no and is planned?

**Comment by Thrameos** [comment, marked as answer]
We support all versions after 9 currently, though the testing is limited to those versions on azure cloud.     If you find a problem with a later version please report it.  17 is a LTS so there is no reason it should not work.

**Reply by Ratgor** [reply]
Thanks! Now it's clear why it not mentioned as 'tested', but if it is maintainable that's ok. (and I have already checked it in my applications, it works ok)

**Reply by Thrameos** [reply]
It was developed using JNI version 1.4 and has a backend that compiles a jar that is compatible with 8/9 and versions going forward.   Thus the only way that it will break is if the security model for Java changes sufficiently that it is unable to use reflection to access the class description/methods/fields.   If they add a concept that requires a new JNI to access a feature then JPype will need to be updated accordingly.    We use a "multi release jar file" so that means that the jar file can mutate depending on the version of Java that is loaded.     The only issue is that as the build process is using the oldest version so that means we end up using reflection to access newer features rather than trying to compile with multiple Java versions (or compile with the latest and try to get it to produce old code.) 

https://www.baeldung.com/java-multi-release-jar

The only time it has broken (other than internal bugs) was in JDK 12 in which they changed the model for assessing caller sensitive methods.   In that version JPype broken because C doesn't have a caller slot and thus all the methods that required the caller to be checked failed (including some of the instrumentation required for reflection).   However, as that was patched it should work going forward at least until they change the security model again.    This was more an oversite on the part of Java developers that JNI protocol does not support specifying a class as the context for a call.
